// ui/src/pages/NewStudy.jsx
import { useState, useRef } from 'react';
import { useNavigate } from 'react-router-dom';
import './NewStudy.css';
import { patientAPI, studyAPI, dicomAPI } from '../services/api';
import { parseDicomMetadata } from '../utils/dicomUtils';

function NewStudy() {
  const navigate = useNavigate();
  const fileInputRef = useRef(null);
  const [formData, setFormData] = useState({
    patientID: '',
    nameSurname: '',
    age: '',
    gender: '',
    clinicalNote: ''
  });
  const [dicomFiles, setDicomFiles] = useState([]);
  const [uploadProgress, setUploadProgress] = useState(0);
  const [progress, setProgress] = useState(0);
  const [isProcessing, setIsProcessing] = useState(false);
  const [studyId, setStudyId] = useState(null);

  const handleInputChange = (e) => {
    setFormData({...formData, [e.target.name]: e.target.value});
  };

  const handleDicomUpload = async (e) => {
    const files = Array.from(e.target.files);
    if (files.length === 0) return;

    // Filter only DICOM files
    const dicomFiles = files.filter(file => 
      file.name.toLowerCase().endsWith('.dcm') || 
      file.name.toLowerCase().endsWith('.dicom') ||
      file.type === 'application/dicom'
    );

    if (dicomFiles.length === 0) {
      alert('Please select valid DICOM files (.dcm or .dicom)');
      return;
    }

    setDicomFiles(dicomFiles);
    
    // Parse first file to extract patient info
    if (dicomFiles.length > 0) {
      try {
        const metadata = await parseDicomMetadata(dicomFiles[0]);
        if (metadata.patientName) {
          setFormData(prev => ({
            ...prev,
            patientID: metadata.patientID || prev.patientID,
            nameSurname: metadata.patientName || prev.nameSurname,
          }));
        }
      } catch (error) {
        console.error('Error parsing DICOM metadata:', error);
      }
    }
  };

  const savePatientAndStudy = async () => {
    try {
      // Validate form
      if (!formData.patientID || !formData.nameSurname) {
        alert('Please fill in Patient ID and Name');
        return null;
      }

      // DICOM files are optional but recommended
      if (dicomFiles.length === 0) {
        const proceed = confirm('No DICOM files selected. Continue without images?\n\nNote: You will not be able to view images in the Review page.');
        if (!proceed) {
          return null;
        }
      }

      // Create patient
      const patientData = {
        patient_id: formData.patientID,
        name: formData.nameSurname,
        age: parseInt(formData.age) || null,
        gender: formData.gender || null
      };

      try {
        console.log('Creating patient:', patientData);
        console.log('API URL:', 'http://localhost:3001/api/patients');
        const response = await patientAPI.create(patientData);
        console.log('Patient created:', response.data);
      } catch (error) {
        // Patient might already exist, that's okay
        console.log('Patient creation error (might already exist):', error.response?.data || error.message);
        console.log('Full error:', error);
        // If error is not "already exists", throw it
        if (error.response?.status !== 500 && !error.message.includes('UNIQUE constraint')) {
          throw error;
        }
      }

      // Create study
      const newStudyId = `STD-${Date.now()}`;
      const studyData = {
        study_id: newStudyId,
        patient_id: formData.patientID,
        study_date: new Date().toISOString().split('T')[0],
        description: formData.clinicalNote || 'CT Chest Study'
      };

      console.log('Creating study:', studyData);
      const studyResponse = await studyAPI.create(studyData);
      console.log('Study created:', studyResponse.data);
      
      setStudyId(newStudyId);

      return newStudyId;
    } catch (error) {
      console.error('Error saving patient/study:', error);
      const errorMsg = error.response?.data?.error || error.message || 'Unknown error';
      alert(`Failed to save patient/study information: ${errorMsg}\n\nPlease check:\n1. Backend server is running (http://localhost:3001)\n2. Patient ID and Name are filled\n3. Browser console for details`);
      return null;
    }
  };

  const startAnalysis = async () => {
    const newStudyId = await savePatientAndStudy();
    if (!newStudyId) return;

    setIsProcessing(true);
    setProgress(0);

    try {
      // Upload DICOM files if available
      setProgress(10);
      if (dicomFiles.length > 0) {
        console.log('Uploading DICOM files...', dicomFiles.length);
        await dicomAPI.uploadFiles(newStudyId, dicomFiles);
        console.log('DICOM files uploaded successfully');
      } else {
        console.log('No DICOM files to upload');
      }
      setProgress(40);

      // Simulate AI analysis
      await new Promise(resolve => setTimeout(resolve, 1000));
      setProgress(70);

      await new Promise(resolve => setTimeout(resolve, 1000));
      setProgress(90);

      // Update study status
      const noduleCount = Math.floor(Math.random() * 5) + 1; // Mock nodule count
      await studyAPI.updateStatus(newStudyId, 'completed', noduleCount);
      
      setProgress(100);
    } catch (error) {
      console.error('Error during analysis:', error);
      alert('Failed to process study. Please try again.');
      setIsProcessing(false);
      setProgress(0);
    }
  };

  const submitResults = () => {
    if (studyId) {
      navigate(`/review/${studyId}`);
    }
  };

  const testBackendConnection = async () => {
    try {
      console.log('Testing backend connection...');
      const response = await fetch('http://localhost:3001/api/health');
      const data = await response.json();
      console.log('Backend response:', data);
      alert(`Backend is working!\n${JSON.stringify(data, null, 2)}`);
    } catch (error) {
      console.error('Backend connection failed:', error);
      alert(`Backend connection failed!\n${error.message}\n\nMake sure backend is running on port 3001`);
    }
  };

  return (
    <div className="worklist">
      <div className="worklist-header">
        <h2>New Study</h2>
        <p>Add a new CT study and start AI analysis (mock)</p>
        <button 
          onClick={testBackendConnection}
          style={{
            marginTop: 8,
            padding: '8px 16px',
            background: '#10b981',
            color: 'white',
            border: 'none',
            borderRadius: 4,
            cursor: 'pointer'
          }}
        >
          🔍 Test Backend Connection
        </button>
      </div>

      <div className="dashboard-section">
        <div className="patient-info-section">
          <div>
            <h3>Patient Information</h3>
            <div style={{display:'grid',gridTemplateColumns:'1fr 1fr',gap:12}}>
              <input 
                name="patientID" 
                placeholder="Patient ID" 
                value={formData.patientID}
                onChange={handleInputChange}
                className="new-study-input"
              />
              <input 
                name="nameSurname" 
                placeholder="Name Surname" 
                value={formData.nameSurname}
                onChange={handleInputChange}
                className="new-study-input"
              />
              <input 
                name="age" 
                placeholder="Age" 
                value={formData.age}
                onChange={handleInputChange}
                className="new-study-input"
              />
              <input 
                name="gender" 
                placeholder="Gender" 
                value={formData.gender}
                onChange={handleInputChange}
                className="new-study-input"
              />
              <input 
                name="clinicalNote" 
                placeholder="Short clinical note" 
                value={formData.clinicalNote}
                onChange={handleInputChange}
                className="new-study-input"
                style={{gridColumn:'1/3'}}
              />
            </div>
          </div>
          <button className="save-btn">Save</button>
        </div>
      </div>

      <div className="dashboard-section">
        <h3>Image Upload</h3>
        <div style={{display:'flex',gap:8, alignItems: 'center'}}>
          <input 
            ref={fileInputRef}
            type="file" 
            accept=".dcm,.dicom,application/dicom"
            multiple
            onChange={handleDicomUpload}
            style={{display: 'none'}}
          />
          <button 
            className="open-btn" 
            onClick={() => fileInputRef.current?.click()}
          >
            📂 Choose DICOM Files ({dicomFiles.length} selected)
          </button>
          <button 
            className="open-btn" 
            onClick={startAnalysis} 
            disabled={isProcessing || dicomFiles.length === 0}
          >
            {isProcessing ? 'Processing...' : 'Start AI Analysis'}
          </button>
        </div>
        
        {dicomFiles.length > 0 && (
          <div style={{marginTop: 12, padding: 12, background: '#1b2430', borderRadius: 8}}>
            <p style={{color: '#64748b', fontSize: '14px'}}>
              ✓ {dicomFiles.length} DICOM file(s) ready for upload
            </p>
          </div>
        )}

        {isProcessing && (
          <div style={{marginTop:16}}>
            <div style={{background:'#1b2430',borderRadius:8,overflow:'hidden'}}>
              <div style={{height:12,width:`${progress}%`,background:'#3b82f6',transition:'width 300ms'}} />
            </div>
            <p style={{marginTop:8}}>
              <strong>Processing:</strong> {progress < 100 ? 'In progress' : 'Process completed'}
            </p>
            {progress === 100 && (
              <button className="submit-btn" onClick={submitResults}>View Results in Review</button>
            )}
          </div>
        )}
      </div>
    </div>
  );
}

export default NewStudy;
